pypeit/spectrographs/keck_kcwi.py

```diff
@@ -1031,46 +1032,164 @@ def get_rawimage(self, raw_file, det):
+            * param[8] = Polynomial scaling coefficient in the spectral direction (coefficient of spec_pixel**3)
+        img : `numpy.ndarray`_
+            Raw image that you want to compute the scattered light model.
+            shape is (nspec, nspat)
+            Model used to calculate the scattered light. This function is used to
+            generate a model of the scattered light, based on a set of model parameters
+            that have been optimized using self.scattered_light().
+        kernel : :obj:`str`_, optional
+            The shape of the kernel to use. The allowed values for KCWI are 'gaussian' and 'lorentzian',
+            but the default value ('gaussian') is known to provide the best fit.
+
+        Returns
+        -------
+        model : `numpy.ndarray`_
+            Model of the scattered light for the input
         """
-        Calculate a model of the scattered light of the input frame.
+        # For clarity, unpack the parameters
+        assert param.size == 9  # For KCWI, the scattered light model requires 9 model parameters
+        sigmx, sigmy, shft_spec, shft_spat, zoom = param[0], param[1], param[2], param[3], param[4]
+        term0, term1, term2, term3 = param[5], param[6], param[7], param[8]
+        # Generate a 2D smoothing kernel
+        if kernel == 'gaussian':
+            # Gaussian
+            subkrnx = np.exp(-0.5 * (np.arange(int(6 * sigmx)) - 3 * sigmx) ** 2 / sigmx ** 2)
+            subkrny = np.exp(-0.5 * (np.arange(int(6 * sigmy)) - 3 * sigmy) ** 2 / sigmy ** 2)
+        elif kernel == 'lorentzian':
+            # Lorentzian
+            subkrnx = sigmx / ((np.arange(int(10 * sigmx)) - 5 * sigmx) ** 2 + sigmx ** 2)
+            subkrny = sigmy / ((np.arange(int(10 * sigmy)) - 5 * sigmy) ** 2 + sigmy ** 2)
+        else:
+            msgs.error(f"Unknown kernel: {kernel}")
+        kernel = np.outer(subkrnx, subkrny)
+        kernel /= np.sum(kernel)
+        # Make a grid of coordinates
+        specvec, spatvec = np.arange(img.shape[0]), np.arange(img.shape[1])
+        spat, spec = np.meshgrid(spatvec, specvec / (specvec.size - 1))
+        # Convolve the input image (note: most of the time is spent here)
+        # oaconvolve is the fastest option when the kernel is much smaller dimensions than the image
+        # scale_img = (term0 + term1 * spec + term2*spec**2 + term3*spec**3) * signal.fftconvolve(img, kernel, mode='same')
+        scale_img = (term0 + term1 * spec + term2 * spec ** 2 + term3 * spec ** 3) * signal.oaconvolve(img, kernel,
+                                                                                                       mode='same')
+        spl = interpolate.RectBivariateSpline(specvec, spatvec, scale_img, kx=1, ky=1)
+        return spl(zoom * (specvec + shft_spec), zoom * (spatvec + shft_spat))
+
+    def scattered_light(self, frame, offslitmask, tbl, detpad=300):
```

Comment:
Similar comment: I would generalize and move much/most of this to `scattlight.py`.

Reply:
See comment above. Perhaps we can discuss this further on Slack, but the reason I've implemented it like this is because I suspect the scattered light model requires spectrograph specific code.